File: src/Python/GeometricObjects/GeometricObjectsDemo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import vtk


def main():
    colors = vtk.vtkNamedColors()

    # Set the background color.
    colors.SetColor("BkgColor", [51, 77, 102, 255])

    # Create container to hold the 3D object generators (sources)
    geometricObjectSources = list()

    # Populate the container with the various object sources to be demonstrated
    geometricObjectSources.append(vtk.vtkArrowSource())
    geometricObjectSources.append(vtk.vtkConeSource())
    geometricObjectSources.append(vtk.vtkCubeSource())
    geometricObjectSources.append(vtk.vtkCylinderSource())
    geometricObjectSources.append(vtk.vtkDiskSource())
    geometricObjectSources.append(vtk.vtkLineSource())
    geometricObjectSources.append(vtk.vtkRegularPolygonSource())
    geometricObjectSources.append(vtk.vtkSphereSource())

    # Create containers for the remaining nodes of each pipeline
    mappers = list()
    actors = list()
    textmappers = list()
    textactors = list()

    # Create a common text property.
    textProperty = vtk.vtkTextProperty()
    textProperty.SetFontSize(16)
    textProperty.SetJustificationToCentered()

    # Create a mapper and actor for each object and the corresponding text label
    for i in range(0, len(geometricObjectSources)):
        geometricObjectSources[i].Update()

        mappers.append(vtk.vtkPolyDataMapper())
        mappers[i].SetInputConnection(geometricObjectSources[i].GetOutputPort())

        actors.append(vtk.vtkActor())
        actors[i].SetMapper(mappers[i])
        actors[i].GetProperty().SetColor(
            colors.GetColor3d("Seashell"))

        textmappers.append(vtk.vtkTextMapper())
        textmappers[i].SetInput(
            geometricObjectSources[i].GetClassName())  # set text label to the name of the object source
        textmappers[i].SetTextProperty(textProperty)

        textactors.append(vtk.vtkActor2D())
        textactors[i].SetMapper(textmappers[i])
        textactors[i].SetPosition(120, 16)  # Note: the position of an Actor2D is specified in display coordinates

    # Define size of the grid that will hold the objects
    gridCols = 3
    gridRows = 3
    # Define side length (in pixels) of each renderer square
    rendererSize = 300

    renderWindow = vtk.vtkRenderWindow()
    renderWindow.SetWindowName("Geometric Objects Demo")
    renderWindow.SetSize(rendererSize * gridCols, rendererSize * gridRows)

    # Set up a grid of viewports for each renderer
    for row in range(0, gridRows):
        for col in range(0, gridCols):
            index = row * gridCols + col

            # Create a renderer for this grid cell
            renderer = vtk.vtkRenderer()
            renderer.SetBackground(colors.GetColor3d("BkgColor"))

            # Set the renderer's viewport dimensions (xmin, ymin, xmax, ymax) within the render window.
            # Note that for the Y values, we need to subtract the row index from gridRows
            # because the viewport Y axis points upwards, but we want to draw the grid from top to down
            viewport = [
                float(col) / gridCols,
                float(gridRows - row - 1) / gridRows,
                float(col + 1) / gridCols,
                float(gridRows - row) / gridRows
            ]
            renderer.SetViewport(viewport)

            # Add the corresponding actor and label for this grid cell, if they exist
            if index < len(geometricObjectSources):
                renderer.AddActor(actors[index])
                renderer.AddActor(textactors[index])
                renderer.ResetCameraClippingRange()

            renderWindow.AddRenderer(renderer)

    interactor = vtk.vtkRenderWindowInteractor()
    interactor.SetRenderWindow(renderWindow)

    renderWindow.Render()
    interactor.Start()


if __name__ == '__main__':
    main()

File: src/CSharp/GeometricObjects/GeometricObjectsDemo.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Windows.Forms;
using System.Diagnostics;
using System.Runtime.InteropServices;

using Kitware.VTK;

namespace ActiViz.Examples {
   public partial class Form1 : Form {
      public Form1() {
         InitializeComponent();
      }


      private void renderWindowControl1_Load(object sender, EventArgs e) {
         try {
            GeometricObjectsDemo();
         }
         catch(Exception ex) {
            MessageBox.Show(ex.Message, "Exception", MessageBoxButtons.OK);
         }
      }


      private void GeometricObjectsDemo() {
         // we create a matrix of 3x3 renderer in our renderwindow
         // each renderer can be interacted with independently from one another
         int rendererSize = 234; // width per renderer
         int gridDimensions = 3;
         this.Size = new System.Drawing.Size(756, 756);
         Random rnd = new Random(2); // for background color variation

         List<vtkPolyDataAlgorithm>  geometricObjectSources = new List<vtkPolyDataAlgorithm>();
         geometricObjectSources.Add(vtkArrowSource.New());
         geometricObjectSources.Add(vtkConeSource.New());
         geometricObjectSources.Add(vtkCubeSource.New());
         geometricObjectSources.Add(vtkCylinderSource.New());
         geometricObjectSources.Add(vtkDiskSource.New());
         geometricObjectSources.Add(vtkLineSource.New());
         geometricObjectSources.Add(vtkRegularPolygonSource.New());
         geometricObjectSources.Add(vtkSphereSource.New());
         geometricObjectSources.Add(vtkEarthSource.New());

         List<vtkRenderer> renderers = new List<vtkRenderer>();
         List<vtkPolyDataMapper> mappers = new List<vtkPolyDataMapper>();
         List<vtkActor> actors = new List<vtkActor>();
         List<vtkTextMapper> textMappers = new List<vtkTextMapper>();
         List<vtkActor2D> textActors = new List<vtkActor2D>();

         // Create one text property for all
         vtkTextProperty textProperty = vtkTextProperty.New();
         textProperty.SetFontSize(18);
         textProperty.SetJustificationToCentered();

         // Create a source, renderer, mapper, and actor
         // for each object 
         for(int i = 0; i < geometricObjectSources.Count; i++) {
            geometricObjectSources[i].Update();
            mappers.Add(vtkPolyDataMapper.New());
            mappers[i].SetInputConnection(geometricObjectSources[i].GetOutputPort());

            actors.Add(vtkActor.New());
            actors[i].SetMapper(mappers[i]);

            textMappers.Add(vtkTextMapper.New());
            textMappers[i].SetInput(geometricObjectSources[i].GetClassName());
            textMappers[i].SetTextProperty(textProperty);

            textActors.Add(vtkActor2D.New());
            textActors[i].SetMapper(textMappers[i]);
            textActors[i].SetPosition(rendererSize/2, 16);

            renderers.Add(vtkRenderer.New());
         }

         // Need a renderer even if there is no actor
         for(int i = geometricObjectSources.Count; i < gridDimensions * gridDimensions; i++) {
            renderers.Add(vtkRenderer.New());
         }

         vtkRenderWindow renderWindow = renderWindowControl1.RenderWindow;
         renderWindow.SetSize(rendererSize*gridDimensions, rendererSize*gridDimensions);

         for(int row = 0; row < gridDimensions; row++) {
            for(int col = 0; col < gridDimensions; col++) {
               int index = row * gridDimensions + col;

               // (xmin, ymin, xmax, ymax)
               double[] viewport = new double[] {
                  (col) * rendererSize / (double)(gridDimensions * rendererSize),
                  (gridDimensions - (row+1)) * rendererSize / (double)(gridDimensions * rendererSize),
                  (col+1)*rendererSize / (double)(gridDimensions * rendererSize),
                  (gridDimensions - row) * rendererSize / (double)(gridDimensions * rendererSize)};

               Debug.WriteLine(viewport[0] + " " + viewport[1] + " " + viewport[2] + " " + viewport[3]);
               renderWindow.AddRenderer(renderers[index]);
               IntPtr pViewport = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(double))*4);
               Marshal.Copy(viewport, 0, pViewport, 4);
               renderers[index].SetViewport(pViewport);
               Marshal.FreeHGlobal(pViewport);
               if(index > geometricObjectSources.Count - 1)
                  continue;

               renderers[index].AddActor(actors[index]);
               renderers[index].AddActor(textActors[index]);
               renderers[index].SetBackground(.2 + rnd.NextDouble() / 8, .3 + rnd.NextDouble() / 8, .4 + rnd.NextDouble() / 8);
            }
         }
      }
   }
}
